extensions/common/context.cc

```diff
@@ -62,11 +62,45 @@ void extractFqdn(const std::string& cluster_name, std::string* fqdn) {
   }
 }
 
-// Extract service name from service fqdn.
-void extractServiceName(const std::string& fqdn, std::string* service_name) {
-  const std::vector<std::string>& parts = absl::StrSplit(fqdn, '.');
-  if (parts.size() > 0) {
-    *service_name = parts[0];
+// Extract service name from service host.
+void extractServiceName(const std::string& host, std::string* service_name,
+                        const std::string& destination_namespace) {
+  // TODO: right now property API only returns cluster with a upstream host, but
+  // not for passthrough or blackhole cluster. Add metrics for these two types
+  // of clusters when property api supports this. if
+  // (!cluster_name.compare(kBlackHoleClusterName) ||
+  //     !cluster_name.compare(kPassThroughClusterName)) {
+  //   *service_name = cluster_name;
+  //   return;
+  // }
+
+  auto name_pos = host.find_first_of(".:");
+  if (name_pos == std::string::npos) {
+    // host is already a short service name. return it directly.
+    *service_name = host;
+    return;
+  }
+  if (host[name_pos] == ':') {
+    // host is `short_service:port`, return short_service name.
+    *service_name = host.substr(0, name_pos);
+    return;
+  }
+
+  auto namespace_pos = host.find_first_of(".:", name_pos + 1);
+  std::string service_namespace = "";
+  if (namespace_pos == std::string::npos) {
+    service_namespace = host.substr(namespace_pos + 1);
```

Comment:
better not to rely on the impl of `std::string::npos` 
how about just `service_namespace = host`?

Reply:
This is actually a bug, `namespace_pos` should be `name_pos`. Not sure I follow the `npos` concern.

Reply:
Got it. There you go.
A typical value of std::string::npos is `-1`. Prior to fix I thought your intention is `service_namespace = host.substr(-1 +1) /* = host.substr(0),  = host */` :)